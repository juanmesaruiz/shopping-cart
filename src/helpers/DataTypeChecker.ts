class DataTypeChecker {
  static isArray(value: unknown): value is Array<unknown> {
    return Array.isArray(value);
  }

  static isArrayOfStrings(value: unknown): value is Array<string> {
    let response = false;
    if (DataTypeChecker.isArray(value)) {
      response = value.every(item => DataTypeChecker.isString(item));
    }
    return response;
  }

  static isString(value: unknown): value is string {
    return typeof value === 'string' || value instanceof String;
  }

  static isNumber(value: unknown): value is number {
    return typeof value === 'number' && isFinite(value);
  }
}

export default DataTypeChecker;
