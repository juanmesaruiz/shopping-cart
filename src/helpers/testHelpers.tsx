import React, { ReactComponentElement } from 'react';

export const MockComponent = (
  name: string,
  props: object
): ReactComponentElement<any> => {
  return (
    <div aria-label={name}>
      <p>{name}</p>
      <p>{JSON.stringify(props).replace(/"/g, "'")}</p>
    </div>
  );
};
