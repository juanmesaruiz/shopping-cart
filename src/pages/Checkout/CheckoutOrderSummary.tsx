import React, { useCallback } from 'react';
import styled from 'styled-components';

import type Checkout from '../../classes/Checkout';
import { setCartItems } from '../../reducers/cartItems';

import { useAppDispatch } from '../../reduxHooks';

const CheckoutOrderSummary = ({ checkout }: { checkout: Checkout }) => {
  const dispatch = useAppDispatch();
  const handlePayButton = useCallback(() => {
    checkout.pay();
    dispatch(setCartItems(0));
  }, [checkout, dispatch]);

  const discounts = checkout.discounts;

  return (
    <CheckoutOrderSummaryContainer>
      <h1 className="main">Order Summary</h1>
      <ul className="summary-items wrapper border">
        <li>
          <span className="summary-items-number">
            {checkout.totalItems} Items
          </span>
          <span className="summary-items-price">
            {checkout.totalWithoutDiscounts()}
            <span className="currency">€</span>
          </span>
        </li>
      </ul>
      <div className="summary-discounts wrapper-half border">
        <h2>Discounts</h2>
        <ul>
          {discounts.length ? (
            discounts.map(discount => (
              <li key={discount.name}>
                <span>{discount.name}</span>
                <span>-{discount.discount}€</span>
              </li>
            ))
          ) : (
            <li>
              <span className="summary-discounts-no-discounts">
                Add products to find some discount for you!
              </span>
            </li>
          )}
        </ul>
      </div>
      <div className="summary-total wrapper">
        <ul>
          <li>
            <span className="summary-total-cost">Total cost</span>
            <span className="summary-total-price">{checkout.total()}€</span>
          </li>
        </ul>
        <button
          onClick={handlePayButton}
          disabled={checkout.totalItems <= 0}
          type="submit"
        >
          Checkout
        </button>
      </div>
    </CheckoutOrderSummaryContainer>
  );
};

const CheckoutOrderSummaryContainer = styled.aside`
  display: flex;
  flex-flow: column wrap;
  padding: 40px 32px;
  width: 312px;
  background-color: #f3f3f3;
  color: #212240;

  > * {
    width: 100%;
  }

  .summary-items-price {
    font-size: 16px;
    font-weight: bold;
  }

  input {
    margin: 8px auto;
    color: #a6a7b3;
    text-indent: 16px;
    text-transform: uppercase;
  }

  input::placeholder {
    color: #a6a7b3;
    text-transform: lowercase;
  }

  ul li {
    display: flex;
    justify-content: space-between;
  }

  ul li > *:nth-child(2) {
    font-weight: bold;
  }

  ul li + li {
    margin-top: 20px;
  }

  .summary-total {
    align-self: flex-end;
    margin-top: auto;
    padding-top: 16px;
    padding-bottom: 0;
    border-top: 1px solid rgba(33, 34, 64, 0.16);
    color: #212240;
  }

  .summary-total-cost {
    flex-basis: 100%;
    text-transform: uppercase;
  }

  .summary-total-price {
    font-weight: bold;
  }

  button[type='submit'] {
    margin-top: 24px;
    padding-top: 16px;
    padding-bottom: 16px;
    width: 100%;
    border-radius: 4px;
    background: #7350ff;
    color: #ffffff;
    font-size: 16px;
    font-weight: bold;
    line-height: 14px;
    cursor: pointer;

    &:disabled {
      background: #a6a7b3;
      cursor: not-allowed;
    }
  }

  .summary-discounts-no-discounts {
    color: #a6a7b3;
    font-size: 12px;
  }

  .wrapper {
    padding: 32px 0;
  }

  .wrapper-half {
    padding: 24px 0;
  }
`;

export default CheckoutOrderSummary;
