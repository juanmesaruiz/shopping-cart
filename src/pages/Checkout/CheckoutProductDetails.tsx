import React, { useCallback, useEffect } from 'react';
import { useSearchParams } from 'react-router-dom';
import styled from 'styled-components';

import type Checkout from '../../classes/Checkout';
import { setCartItems } from '../../reducers/cartItems';

import { QUERY_PARAM_PRODUCT_DETAIL } from '../../config/constants';
import {
  findProduct,
  getProduct,
  getProductCatalog,
} from '../../helpers/productsHelper';
import { useAppDispatch } from '../../reduxHooks';

const CheckoutProductDetails = ({ checkout }: { checkout: Checkout }) => {
  const dispatch = useAppDispatch();
  const [searchParams, setSearchParams] = useSearchParams();
  const productDetailCode = searchParams.get(QUERY_PARAM_PRODUCT_DETAIL);

  const removeCodeFromSearchParams = useCallback(
    () => setSearchParams({}),
    [setSearchParams]
  );

  const handlePressedKey = useCallback(
    (event: KeyboardEvent) => {
      if (event.key === 'Escape') {
        removeCodeFromSearchParams();
      }
    },
    [removeCodeFromSearchParams]
  );

  useEffect(() => {
    if (productDetailCode) {
      document.addEventListener('keyup', handlePressedKey);
    } else {
      document.removeEventListener('keyup', handlePressedKey);
    }
  }, [productDetailCode, handlePressedKey]);

  if (productDetailCode) {
    const product = getProduct(getProductCatalog(), productDetailCode);
    const {
      code,
      description,
      imgAlt,
      imgDetailSrc,
      name,
      price,
      productCode,
      stock,
    } = product;

    const cartItem = findProduct(checkout.cartItems, code);
    const productAmount = cartItem ? cartItem.amount : 0;
    const isDisabledAddToCart = productAmount >= stock;

    const handleAddToCart = () => {
      const amount = +1;
      checkout.scan(code, amount);
      dispatch(setCartItems(amount));
    };

    return (
      <CheckoutProductDetailsContainer>
        <figure className="productDetails-image">
          <img src={imgDetailSrc} alt={imgAlt} />
        </figure>
        <aside className="productDetails-product-details">
          <div className="productDetails-close">
            <button onClick={removeCodeFromSearchParams}>X</button>
          </div>
          <div className="productDetails-name">
            <span>{name}</span>
            <span>{price} â‚¬</span>
          </div>
          <div className="productDetails-description">{description}</div>
          <div className="productDetails-ref">Product code {productCode}</div>
          <button
            onClick={handleAddToCart}
            disabled={isDisabledAddToCart}
            className="productDetails-button"
          >
            Add to cart
          </button>
          {isDisabledAddToCart && (
            <span className="no-stock">
              Sorry, there are no more units available
            </span>
          )}
        </aside>
      </CheckoutProductDetailsContainer>
    );
  }

  return null;
};

const CheckoutProductDetailsContainer = styled.div`
  background: white;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  display: flex;

  .productDetails-image {
    align-self: center;
    text-align: center;
    width: 826px;

    img {
      max-height: 648px;
      vertical-align: middle;
    }
  }

  .productDetails-product-details {
    width: calc(100% - 826px);
    padding: 30px 20px;

    .productDetails-close {
      font-size: 20px;
      text-align: right;

      button {
        background: none;
        cursor: pointer;
      }
    }

    .productDetails-name {
      border-bottom: 1px solid #a6a7b3;
      display: flex;
      font-size: 20px;
      font-weight: bold;
      justify-content: space-between;
      margin-top: 50px;
      padding-bottom: 15px;
    }

    .productDetails-description {
      border-bottom: 1px solid #a6a7b3;
      color: #717285;
      font-size: 16px;
      height: 250px;
      line-height: 1.5;
      padding: 30px 0;
    }

    .productDetails-ref {
      color: #a6a7b3;
      font-size: 14px;
      padding: 30px 0;
    }

    .productDetails-button {
      margin-top: 24px;
      padding-top: 16px;
      padding-bottom: 16px;
      width: 100%;
      border-radius: 4px;
      background: #7350ff;
      color: #ffffff;
      font-size: 16px;
      font-weight: bold;
      line-height: 14px;
      cursor: pointer;

      &:disabled {
        background: #a6a7b3;
        cursor: not-allowed;
      }
    }
  }
`;

export default CheckoutProductDetails;
