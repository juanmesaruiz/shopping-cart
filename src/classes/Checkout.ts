import { getDiscountPerProduct } from '../helpers/discountsHelper';
import {
  findProduct,
  getProduct,
  getProductCatalog,
} from '../helpers/productsHelper';

class Checkout {
  cartItems: Array<{ code: string; amount: number }> = [];
  discounts: Array<{ discount: number; name: string }> = [];
  totalItems: number = 0;

  private setDiscounts(productCode: string, totalAmount: number) {
    const currentProduct = getProduct(getProductCatalog(), productCode);

    this.discounts = getDiscountPerProduct({
      product: currentProduct,
      productAmount: totalAmount,
      currentDiscounts: this.discounts,
    });
  }

  scan(code: string, amount = +1): this {
    const currentCartItem = findProduct(this.cartItems, code);
    let totalAmount = 0;

    if (currentCartItem) {
      totalAmount = Number(currentCartItem.amount) + amount;
      currentCartItem.amount = totalAmount;
    } else if (amount > 0) {
      totalAmount = amount;
      this.cartItems = [...this.cartItems, { code, amount }];
    }

    this.setDiscounts(code, totalAmount);

    this.totalItems = this.totalItems + amount;

    return this;
  }

  setProductAmount(code: string, totalAmount: number): this {
    const productData = getProduct(getProductCatalog(), code);
    const currentCartItem = findProduct(this.cartItems, code);

    const newTotalAmount =
      totalAmount > productData.stock ? productData.stock : totalAmount;
    let newItemAmountOnCart = newTotalAmount;

    if (currentCartItem) {
      newItemAmountOnCart = newTotalAmount - currentCartItem.amount;
      currentCartItem.amount = newTotalAmount;
    } else if (newTotalAmount > 0) {
      this.cartItems = [...this.cartItems, { code, amount: newTotalAmount }];
    }

    this.totalItems = this.totalItems + newItemAmountOnCart;

    this.setDiscounts(code, totalAmount);

    return this;
  }

  totalWithoutDiscounts(): number {
    const initialAmount = 0;

    if (this.cartItems.length) {
      return this.cartItems.reduce((totalAmount, { code, amount }) => {
        const currentProduct = getProduct(getProductCatalog(), code);

        return totalAmount + amount * currentProduct?.price;
      }, initialAmount);
    }

    return initialAmount;
  }

  total(): number {
    const totalWithoutDiscounts = this.totalWithoutDiscounts();
    const discounts = Object.values(this.discounts).reduce(
      (totalAmount, { discount }) => totalAmount + discount,
      0
    );

    return totalWithoutDiscounts - discounts;
  }

  pay(): this {
    this.cartItems = [];
    this.discounts = [];
    this.totalItems = 0;

    return this;
  }
}

export default Checkout;
