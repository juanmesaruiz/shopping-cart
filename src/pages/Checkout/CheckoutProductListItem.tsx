import React, { ChangeEvent, useCallback } from 'react';
import { useSearchParams } from 'react-router-dom';

import type Checkout from '../../classes/Checkout';
import type Product from '../../models/Product';
import { setCartItems } from '../../reducers/cartItems';

import { QUERY_PARAM_PRODUCT_DETAIL } from '../../config/constants';
import { findProduct } from '../../helpers/productsHelper';
import { useAppDispatch } from '../../reduxHooks';

const CheckoutProductListItem = ({
  checkout,
  product: { code, imgAlt, imgCatalogSrc, name, price, productCode, stock },
}: {
  checkout: Checkout;
  product: Product;
}) => {
  const [, setSearchParams] = useSearchParams();
  const cartItem = findProduct(checkout.cartItems, code);
  const productAmount = cartItem ? cartItem.amount : 0;

  const dispatch = useAppDispatch();

  const handleAmountClick = useCallback(
    (amount: number) => {
      checkout.scan(code, amount);
      dispatch(setCartItems(checkout.totalItems));
    },
    [checkout, code, dispatch]
  );

  const handleNameClick = useCallback(
    () => setSearchParams({ [QUERY_PARAM_PRODUCT_DETAIL]: code }),
    [code, setSearchParams]
  );

  const handleInputChange = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      const newItemStock = parseInt(event.target.value);

      if (newItemStock !== productAmount) {
        checkout.setProductAmount(code, newItemStock ? newItemStock : 0);
        dispatch(setCartItems(checkout.totalItems));
      }
    },
    [checkout, code, dispatch, productAmount]
  );

  return (
    <li className="product row">
      <div className="col-product">
        <figure className="product-image">
          <img src={imgCatalogSrc} alt={imgAlt} />
          <div className="product-description">
            <h1 onClick={handleNameClick}>{name}</h1>
            <p className="product-code">Product code {productCode}</p>
          </div>
        </figure>
      </div>
      <div className="col-quantity">
        <button
          disabled={stock === 0 || productAmount === 0}
          className="count"
          onClick={() => handleAmountClick(-1)}
        >
          -
        </button>
        <input
          aria-label="input-amount"
          disabled={stock <= 0}
          type="number"
          className="product-quantity"
          onChange={handleInputChange}
          value={productAmount.toString()}
          min={0}
          max={stock}
        />
        <button
          disabled={productAmount >= stock}
          className="count"
          onClick={() => handleAmountClick(+1)}
        >
          +
        </button>
        {productAmount >= stock && (
          <span className="no-stock">
            Sorry, there are no more units available
          </span>
        )}
      </div>
      <div className="col-price">
        <span className="product-price">{price}</span>
        <span className="product-currency currency">€</span>
      </div>
      <div className="col-total">
        <span className="product-price">{productAmount * price}</span>
        <span className="product-currency currency">€</span>
      </div>
    </li>
  );
};

export default CheckoutProductListItem;
