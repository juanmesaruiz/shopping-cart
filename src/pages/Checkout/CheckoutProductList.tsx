import React from 'react';
import styled from 'styled-components';

import type Checkout from '../../classes/Checkout';

import CheckoutProductListItem from './CheckoutProductListItem';
import { getProductCatalog } from '../../helpers/productsHelper';

const CheckoutProductList = ({ checkout }: { checkout: Checkout }) => {
  const products = getProductCatalog();

  return (
    <CheckoutProductListContainer>
      <h1 className="main">Shopping cart</h1>
      <ul className="products-list tableHead">
        <li className="products-list-title row">
          <div className="col-product">Product details</div>
          <div className="col-quantity">Quantity</div>
          <div className="col-price">Price</div>
          <div className="col-total">Total</div>
        </li>
      </ul>
      <ul className="products-list">
        {products.map(product => (
          <CheckoutProductListItem
            checkout={checkout}
            key={product.code}
            product={product}
          />
        ))}
      </ul>
    </CheckoutProductListContainer>
  );
};

const CheckoutProductListContainer = styled.section`
  flex: 1;
  padding: 40px 32px 40px 56px;

  .tableHead {
    padding: 32px 0;
  }

  [class*='col-'] {
    display: flex;
    align-items: center;
  }

  [class*='col-']:nth-child(n + 2) {
    justify-content: center;
  }

  .col-product {
    width: 45%;
  }

  .col-quantity {
    flex-flow: row wrap;
    width: 20%;
  }

  .col-price {
    width: 20%;
  }

  .col-total {
    width: 15%;
  }

  input {
    width: 100%;
    height: 48px;
    border-radius: 4px;
  }

  button.count {
    padding: 0 8px;
    height: 40px;
    border: none;
    background: transparent;
    color: #7350ff;
    font-weight: bold;
    cursor: pointer;
  }

  button.count:disabled {
    color: #a6a7b3;
    cursor: not-allowed;
  }

  .products-list {
    position: relative;
    width: 100%;
  }

  .product-image {
    display: flex;
    align-items: center;
    flex-flow: row nowrap;
  }

  .product-image img {
    margin-right: 16px;
    width: 72px;
    height: 72px;
    border: 1px solid #cacad1;
    border-radius: 4px;
  }

  .product-code {
    border-radius: 4px;
    color: #a6a7b3;
    letter-spacing: 0.13px;
    font-weight: 400;
  }

  .product-quantity {
    width: 40px;
    height: 40px;
    border: 2px solid #dbdbe0;
    border-radius: 4px;
    text-align: center;
  }

  .product-price {
    color: #000000;
    font-size: 16px;
  }

  .product .clear {
    border: none;
    background: transparent;
  }

  .product .clear span {
    position: relative;
    display: block;
    width: 16px;
    height: 16px;
  }

  .product .clear span:before,
  .product .clear span:after {
    content: '';
    position: absolute;
    top: 8px;
    left: 2px;
    display: block;
    width: 12px;
    height: 2px;
    border-radius: 8px;
    background-color: #a6a7b3;
    transform: rotate(-45deg);
    transform-origin: center center;
  }

  .product .clear span:before {
    transform: rotate(45deg);
  }

  .product h1 {
    color: #7350ff;
    cursor: pointer;
  }

  .row {
    position: relative;
    display: flex;
    flex-flow: row nowrap;
    justify-content: space-between;
  }

  .row + .row {
    margin-top: 32px;
  }
`;

export default CheckoutProductList;
