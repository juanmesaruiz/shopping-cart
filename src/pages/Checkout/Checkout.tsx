import React from 'react';
import styled from 'styled-components';

import type TypeCheckout from '../../classes/Checkout';

import CheckoutOrderSummary from './CheckoutOrderSummary';
import CheckoutProductDetails from './CheckoutProductDetails';
import CheckoutProductList from './CheckoutProductList';
import { useAppSelector } from '../../reduxHooks';

const Checkout = ({ checkout }: { checkout: TypeCheckout }) => {
  useAppSelector(state => state.cartItems.amount);

  return (
    <CheckoutContainer>
      <CheckoutProductList checkout={checkout} />
      <CheckoutOrderSummary checkout={checkout} />
      <CheckoutProductDetails checkout={checkout} />
    </CheckoutContainer>
  );
};

const CheckoutContainer = styled.main`
  background-color: #ffffff;
  border-radius: 4px;
  display: flex;
  height: calc(100% - 64px);
  left: 50%;
  max-height: 648px;
  max-width: 1088px;
  overflow-x: hidden;
  overflow-y: auto;
  position: fixed;
  top: 50%;
  transform: translate(-50%, -50%);
  width: calc(100% - 64px);
`;

export default Checkout;
