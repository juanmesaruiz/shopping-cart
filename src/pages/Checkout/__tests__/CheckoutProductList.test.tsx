import React from 'react';
import { render } from '@testing-library/react';

import CheckoutProductList from '../CheckoutProductList';
import Checkout from '../../../classes/Checkout';
import { MockComponent } from '../../../helpers/testHelpers';

jest.mock(
  '../CheckoutProductListItem',
  () => (props: object) => MockComponent('CheckoutProductListItem', props)
);

describe('CheckoutProductList', () => {
  const checkout = new Checkout();

  const defaultProps = {
    checkout,
  };
  const getComponent = (props: { checkout: Checkout } = defaultProps) => (
    <CheckoutProductList {...props} />
  );

  const setup = () => render(getComponent());

  it('Should match snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });
});
