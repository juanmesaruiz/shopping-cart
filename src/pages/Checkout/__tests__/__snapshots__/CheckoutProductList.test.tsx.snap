// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutProductList Should match snapshot 1`] = `
.c0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  padding: 40px 32px 40px 56px;
}

.c0 .tableHead {
  padding: 32px 0;
}

.c0 [class*='col-'] {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 [class*='col-']:nth-child(n + 2) {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0 .col-product {
  width: 45%;
}

.c0 .col-quantity {
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  width: 20%;
}

.c0 .col-price {
  width: 20%;
}

.c0 .col-total {
  width: 15%;
}

.c0 input {
  width: 100%;
  height: 48px;
  border-radius: 4px;
}

.c0 button.count {
  padding: 0 8px;
  height: 40px;
  border: none;
  background: transparent;
  color: #7350ff;
  font-weight: bold;
  cursor: pointer;
}

.c0 button.count:disabled {
  color: #a6a7b3;
  cursor: not-allowed;
}

.c0 .products-list {
  position: relative;
  width: 100%;
}

.c0 .product-image {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
}

.c0 .product-image img {
  margin-right: 16px;
  width: 72px;
  height: 72px;
  border: 1px solid #cacad1;
  border-radius: 4px;
}

.c0 .product-code {
  border-radius: 4px;
  color: #a6a7b3;
  -webkit-letter-spacing: 0.13px;
  -moz-letter-spacing: 0.13px;
  -ms-letter-spacing: 0.13px;
  letter-spacing: 0.13px;
  font-weight: 400;
}

.c0 .product-quantity {
  width: 40px;
  height: 40px;
  border: 2px solid #dbdbe0;
  border-radius: 4px;
  text-align: center;
}

.c0 .product-price {
  color: #000000;
  font-size: 16px;
}

.c0 .product .clear {
  border: none;
  background: transparent;
}

.c0 .product .clear span {
  position: relative;
  display: block;
  width: 16px;
  height: 16px;
}

.c0 .product .clear span:before,
.c0 .product .clear span:after {
  content: '';
  position: absolute;
  top: 8px;
  left: 2px;
  display: block;
  width: 12px;
  height: 2px;
  border-radius: 8px;
  background-color: #a6a7b3;
  -webkit-transform: rotate(-45deg);
  -ms-transform: rotate(-45deg);
  transform: rotate(-45deg);
  -webkit-transform-origin: center center;
  -ms-transform-origin: center center;
  transform-origin: center center;
}

.c0 .product .clear span:before {
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
}

.c0 .product h1 {
  color: #7350ff;
  cursor: pointer;
}

.c0 .row {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c0 .row + .row {
  margin-top: 32px;
}

<div>
  <section
    class="c0"
  >
    <h1
      class="main"
    >
      Shopping cart
    </h1>
    <ul
      class="products-list tableHead"
    >
      <li
        class="products-list-title row"
      >
        <div
          class="col-product"
        >
          Product details
        </div>
        <div
          class="col-quantity"
        >
          Quantity
        </div>
        <div
          class="col-price"
        >
          Price
        </div>
        <div
          class="col-total"
        >
          Total
        </div>
      </li>
    </ul>
    <ul
      class="products-list"
    >
      <div
        aria-label="CheckoutProductListItem"
      >
        <p>
          CheckoutProductListItem
        </p>
        <p>
          {'checkout':{'cartItems':[],'discounts':[],'totalItems':0},'product':{'_code':'TSHIRT','_description':'Online Exclusive. Regular fit. Recycled viscose blend fabric. Lightweight fabric. Round neck.\\n\\nThese products have been made with fibers and/or sustainable production processes, thus reducing their environmental impact.','_imgAlt':'T-shirt image','_imgCatalogSrc':'assets/imgCatalog/tshirt.png','_imgDetailSrc':'assets/imgDetail/tshirt.jpg','_name':'T-Shirt','_price':20,'_productCode':'X7R2OPX','_stock':99}}
        </p>
      </div>
      <div
        aria-label="CheckoutProductListItem"
      >
        <p>
          CheckoutProductListItem
        </p>
        <p>
          {'checkout':{'cartItems':[],'discounts':[],'totalItems':0},'product':{'_code':'MUG','_description':'Pantone 7664. Official color.','_imgAlt':'Mug image','_imgCatalogSrc':'assets/imgCatalog/mug.png','_imgDetailSrc':'assets/imgDetail/mug.jpg','_name':'Coffee Mug','_price':5,'_productCode':'X2G2OPZ','_stock':99}}
        </p>
      </div>
      <div
        aria-label="CheckoutProductListItem"
      >
        <p>
          CheckoutProductListItem
        </p>
        <p>
          {'checkout':{'cartItems':[],'discounts':[],'totalItems':0},'product':{'_code':'CAP','_description':'Baseball type. Nylon. Buckle closure.','_imgAlt':'Cap image','_imgCatalogSrc':'assets/imgCatalog/cap.png','_imgDetailSrc':'assets/imgDetail/cap.jpg','_name':'Cap','_price':10,'_productCode':'X3W2OPY','_stock':99}}
        </p>
      </div>
      <div
        aria-label="CheckoutProductListItem"
      >
        <p>
          CheckoutProductListItem
        </p>
        <p>
          {'checkout':{'cartItems':[],'discounts':[],'totalItems':0},'product':{'_code':'WATER','_description':'Old plastic bottle. All our new bottles are made of cardboard, they will contain 330 ml of water, as well as 70% renewable raw material.','_imgAlt':'Water bottle image','_imgCatalogSrc':'assets/imgCatalog/water.png','_imgDetailSrc':'assets/imgDetail/water.jpg','_name':'Water','_price':0.5,'_productCode':'X1M46NZ','_stock':0}}
        </p>
      </div>
    </ul>
  </section>
</div>
`;
